test_partner_client.py

```diff
@@ -0,0 +1,131 @@
+
+        valid_covid_responses = ["negative", "positive", "failed"]
+
+        if cov19_result not in valid_covid_responses:
+            raise AssertionError("cov_19_result has to have one of the following values: {}".format(
+                ", ".join(valid_covid_responses)))
+
+    def get_as_dict(self):
+        return vars(self)
+
+
+class FailedInContactingTestPartner(Exception):
+    pass
+
+
+class TestPartnerClient(object):
+    # This is valid for v.6 for the parters API
+
+    def __init__(self, config):
+        # TODO Figure out how to configure this, given Claritys normal setup
+        super().__init__()
+        self._url = config.get("test_partner_url")
+        self._user = config.get("test_partner_user")
+        self._password = config.get("test_partner_password")
+
+    # TODO decide if we want to retry or not. It depends a little bit on the
+    #      rest of the scenario. This could potentially take us into a long chain
+    #      of retries. That would mean things take a long time before failing.
+    #      Don't know what is best here.
+    @retry((ConnectionError, Timeout), tries=3, delay=2, backoff=2)
+    def send_single_sample_result(self, test_partner_sample_info):
+
+        if not isinstance(TestPartnerSampleInformation):
+            raise AssertionError(
+                "Expected type of test_partner_sample_info is TestPartnerSampleInformation")
+
+        parameters = test_partner_sample_info.get_as_dict().update(
+            {"user": self._user, "password": self._password})
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+        response = requests.post(
+            self._url, header=headers, data=parameters)
+
+        if not response.status_code == 200:
+            raise FailedInContactingTestPartner("Did not get a 200 response from test partner. " +
+                                                "Response status code was: {}" +
+                                                "and response text: {}".format(response.status_code, response.text))
+
+        # Example of successful API call response
+        # success
+        # referrral_code=1234567890
+        # covid19_result=failed
+        # arrival_date=20140310
+        # test_date=20140310
+        # result_date=20140310
+        # comment=ett testmeddelande
+        response_text_parts = response.text.split("\n")
+        if response_text_parts[0] == "success":
+            # REVIEW Should we for a True/False return for this, or something else?
+            #        Since it also raises on some failures, it might not be necessary,
+            #        if we don't expect to do anything with the results.
+            return True
+        else:
+            return False
+
+        # TODO Figure out how text response should be parsed if there is an error.
+
+    # TODO Finish this when/if there is a batch API at the partner.
+    @retry((ConnectionError, Timeout), tries=3, delay=2, backoff=2)
```

Comment:
Yeah on this topic. We should perhaps take up the discussion about having this run as a service or not. I wonder how it will be for the workflow if the lab technician will have to wait (e.g.) 96x1sec to close a step. It could even take longer time than that, because I would want to update the lims after each call, so we don't lose the information on successful runs if the 95th sample fails.

Reply:
I talked to them today about providing a batch api, and they seemed positive. So I think if we can get that we don't have to deal with the additional complexity of the service. If not, we will figure something out.